Add rescript-mode recipe

jjlee opened the pull request:
### Brief summary of what the package does

This is a new major mode for the ReScript language (https://rescript-lang.org/).

### Direct link to the package repository

https://github.com/jjlee/rescript-mode

### Your association with the package

I am the maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

Re checkdoc and the single-file advice:

The indentation code in rescript-indent.el is derived from js.el in Emacs (it's just that code cut down, currently), so I've kept that in a separate file -- it does not pass checkdoc, but it probably wouldn't benefit anybody for me to go filling in docstrings for a lot of details that I'll get wrong because I don't know the code in that much detail (not yet anyway).  I don't think rescript-indent.el is much use separate from rescript-mode.el, but it's pretty separate from it in implementation, so because of that and the checkdoc issue I kept them as two files despite the advice in the MELPA contributors guide.

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings (see caveat above)
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented on recipes/rescript-mode:
```diff
@@ -0,0 +1,4 @@
+(rescript-mode
+ :fetcher github
+ :repo "jjlee/rescript-mode"
+ :branch "master")
```
```suggestion
 :repo "jjlee/rescript-mode")
```
Not necessary, as long as this is your default branch (we prefer MELPA to build from the default branch)

jjlee replied:
I read that in CONTRIBUTORS.org and didn't understand "the default branch"... but I guess this is referring to github having a concept of default branch?

riscy reviewed (commented):
> The indentation code in rescript-indent.el is derived from js.el in Emacs (it's just that code cut down, currently), so I've kept that in a separate file

Okay, in this case I think it makes sense to grandfather in the checkdoc state of that code.  There are some small improvements you could make when you get around to it:

- rescript-indent.el#L206: Could this point/point-min comparison use `bobp`?
- rescript-indent.el#L353: Could this point/point-min comparison use `bobp`?
- rescript-indent.el#L387: Could this point/point-min comparison use `bobp`?
- rescript-indent.el#L458: Optionally use `unless ...` instead of `when (not ...)`

as well as a couple of package-lints (using version 20210425.3):
```
2 issues found:
315:9: warning: Closing parens should not be wrapped onto new lines.
555:14: error: You should depend on (emacs "26.1") if you need `prog-first-column'.
```

And speaking of dependencies, I suggest removing rescript-mode-pkg.el. MELPA will automatically create a pkg.el file from the contents of your main file (rescript-mode.el), and you'll only have to list your dependencies once, on the Package-Requires line.

Lastly a couple more on rescript-mode.el:
- rescript-mode.el#L165: It's safer to sharp-quote function names; use `#'`
- rescript-mode.el#L184: Terminate auto-mode-alist entries with `\\'`

jjlee commented:
A question: I notice that I've actually been perhaps a bit sloppy with naming:

Starting from js.el, I renamed everything from `js-*` to `rescript-*`, I guess strictly I should rename to `rescript-mode-*`.  And starting from `reason-mode.el`, I renamed `reason-*` to `rescript-*`, resulting in (as for `reason-mode`) some names like `rescript-re-indent` and others like `rescript-mode-consts` (I think I can also make these private).

So:
1. Presumably I should rename everything to start with "reason-mode-"?
2. Does that include the customize group?  I couldn't find anything about customize naming conventions...

jjlee commented:
One more question: I now have the latest package-lint again (package-lint-20210425.3), but I don't see for example the `\\'` issue show up when I run `package-lint-current-buffer` on `rescript-mode.el` when I deliberately add that issue back in my working copy by changing the `\\'` back to '$' (on top of 074a2d0).  I wonder what I'm doing differently than you?

purcell commented:
>     1. Presumably I should rename everything to start with "rescript-mode-"?  On the other hand I see for example `rust-mode.el` uses just `rust-` as the prefix, maybe that's conventional with language major modes?

Yes, since this is a major mode, you should be able to get away with `rescript-` as the prefix.

>     2\. Does that include the customize group?  I couldn't find anything about customize naming conventions...

`package-lint` will actually warn you that the custom group should be called `rescript` rather than `rescript-mode`: the former is more conventional.

> I don't see for example the `\\'` issue show up when I run `package-lint-current-buffer` on `rescript-mode.el` when I deliberately add that issue back in my working copy by changing the `\\'` back to '$' (on top of 074a2d0). I wonder what I'm doing differently than you?

Chris uses some additional elisp linting scripts on top of package-lint, to cover common issues which aren't necessarily package problems.

purcell commented:
Merged now in any case, thanks! I sent a quick PR with a small fix.

riscy commented:
> One more question: I now have the latest package-lint again (package-lint-20210425.3), but I don't see for example the \\' issue show up

That's a check I sneak in as part of a bundle of checks I run [here](https://github.com/riscy/melpazoid).  Some of the more useful/stable checks I should look at PR'ing against package-lint. :)
